File: otsu.py
import matplotlib.pyplot as plt
from skimage import io
from skimage.filters import threshold_otsu
import numpy as np

# Load 16-bit grayscale image (replace with your file path)
# filename = 'test_image.tif'  # Supports TIFF, PNG - use .tif for 16-bit
filename = 'img\\Circle.tif'  # Supports TIFF, PNG - use .tif for 16-bit
image = io.imread(filename, as_gray=True)

# Ensure it's 16-bit uint16
print(f"Image loaded: {filename}")
print(f"Shape: {image.shape}, Dtype: {image.dtype}, Range: [{image.min():.0f}, {image.max():.0f}]")

# Compute Otsu threshold
thresh = threshold_otsu(image)
print(f"Otsu threshold value: {thresh:.2f}")

# Binarize
binary = image > thresh

# Plot results
fig, axes = plt.subplots(ncols=3, figsize=(10, 3.5))
ax = axes.ravel()

ax[0].imshow(image, cmap=plt.cm.gray)
ax[0].set_title('Original 16-bit')
ax[0].axis('off')

ax[1].hist(image.ravel(), bins=1024)  # More bins for 16-bit data
ax[1].set_title('Histogram')
ax[1].axvline(thresh, color='r', linewidth=2)
ax[1].set_xlim(0, image.max())

ax[2].imshow(binary, cmap=plt.cm.gray)
ax[2].set_title(f'Thresholded (T={thresh:.0f})')
ax[2].axis('off')

plt.tight_layout()
plt.show()

File: otsu.m
% Script to load 16-bit grayscale image and compute Otsu threshold with otsuthresh

% Specify your 16-bit grayscale image file
% filename = 'test_image.tif';  % Replace with your file path (TIFF/PNG)
filename = 'img\Zippo.tif';  % Replace with your file path (TIFF/PNG)

% Read 16-bit image (uint16)
img = imread(filename);

% Verify grayscale
if size(img, 3) > 1
    error('Image must be grayscale.');
end

% Display image info
fprintf('Image: %s\nSize: %dx%d\nClass: %s\nRange: [%d, %d]\n', ...
    filename, size(img,1), size(img,2), class(img), ...
    double(min(img(:))), double(max(img(:))));

% Compute histogram (use 65536 bins for full 16-bit precision, or fewer for speed)
num_bins = 65536;  % Full 16-bit range; use 1024 or 4096 for faster computation
[counts, ~] = imhist(img, num_bins);

% Compute Otsu threshold from histogram counts (returns [0,1])
T_normalized = otsuthresh(counts);

% Convert to absolute threshold value
% max_val = double(max(img(:)));
max_val = 65536.0;

T_absolute = T_normalized * (max_val);  % Scale to image range

fprintf('Otsu threshold (normalized): %.4f\n', T_normalized);
fprintf('Otsu threshold (absolute): %.0f\n', T_absolute);

% Optional: Binarize and display results
BW = imbinarize(img, T_normalized);
figure;
subplot(1,2,1); imshow(img, []); title('Original 16-bit');
subplot(1,2,2); imshow(BW); title(sprintf('Binarized (T=%.4f)', T_normalized));
